### PR TITLE
Numpy Ops Large Tensor Tests

### PR DESCRIPTION
**Pull request by Zha0q1**
This PR adds large tensor tests to numpy operators.

This is meant for progress sharing and not for merging.

To test run `nosetests --logging-level=DEBUG --verbose -s tests/nightly/test_np_large_array.py:{test_name}`

--------------------
Update: tested count: 49 numpy 10 npx

### CODE REVIEW

**Comment by access2rohit** [comment]
@Zha0q1 paste the output using pytest run and not nosetests. Get these ops merged. You can add new ops to a different PR.

**Comment by access2rohit** [comment]
Use @mxnet-label-bot to mark your PR as WIP/ready for review/ready to merge

**Comment by Zha0q1** [comment]
Let me add in backward tests and also skip the ones that we know will fail. Aiming for mergeing tomorrow

File: tests/nightly/test_np_large_array.py [hunk 1]
```diff
@@ -76,3 +78,459 @@ def test_softmax():
...
+
+@use_np
+def test_ones():
+    A = np.ones((INT_OVERFLOW, 2))
+    assert A.shape == (INT_OVERFLOW, 2)
+
+@use_np
+def test_zeros():
+    A = np.zeros((INT_OVERFLOW, 2))
+    assert A.shape == (INT_OVERFLOW, 2)
```
Comment (access2rohit) [inline tests/nightly/test_np_large_array.py:97]: Can you also add checks for single value at [0] or [-1]

File: tests/nightly/test_np_large_array.py [hunk 2]
```diff
@@ -76,3 +78,459 @@ def test_softmax():
...
+  _ _ _  _ _ __  _ __ _  _
+ | ' \ || | '  \| '_ \ || |
+ |_||_\_,_|_|_|_| .__/\_, |
+                |_|   |__/
+'''
+
+@use_np
+def test_ones():
+    A = np.ones((INT_OVERFLOW, 2))
+    assert A.shape == (INT_OVERFLOW, 2)
```
Comment (access2rohit) [inline tests/nightly/test_np_large_array.py:92]: values checks

File: tests/nightly/test_np_large_array.py [hunk 3]
```diff
@@ -76,3 +78,459 @@ def test_softmax():
...
+    assert A.grad.shape == (INT_OVERFLOW, 2)
+
+@use_np
+def test_absolute():
+    A = np.ones((INT_OVERFLOW, 2))
+    A.attach_grad()
+    with mx.autograd.record():
+        B = np.absolute(A)
+    print(B)
+    assert B.shape == (INT_OVERFLOW, 2)
```
Comment (access2rohit) [inline tests/nightly/test_np_large_array.py:117]: same as above

File: tests/nightly/test_np_large_array.py [hunk 4]
```diff
@@ -76,3 +78,459 @@ def test_softmax():
...
+
+@use_np
+def test_add():
+    A = np.ones((INT_OVERFLOW, 2))
+    B = np.ones((INT_OVERFLOW, 2))
+    A.attach_grad()
+    with mx.autograd.record():
+        C = np.add(A, B)
+    print(C)
+    assert C.shape == (INT_OVERFLOW, 2)
```
Comment (access2rohit) [inline tests/nightly/test_np_large_array.py:129]: same as above

File: tests/nightly/test_np_large_array.py [hunk 5]
```diff
@@ -76,3 +78,459 @@ def test_softmax():
...
+
+# this will fail; broadcast needs to be fixed
+# TODO add backward test after forward is fixed
+@use_np
+@pytest.mark.skip(reason='Does not support large tensor; to be fixed')
+def test_add_broadcast():
+    A = np.ones((INT_OVERFLOW, 2))
+    B = np.ones((INT_OVERFLOW, 1))
+    C = np.add(A, B)
+    print(C)
```
Comment (access2rohit) [inline tests/nightly/test_np_large_array.py:141]: don't print. add assert checks for single value

File: tests/nightly/test_np_large_array.py [hunk 6]
```diff
@@ -76,3 +78,459 @@ def test_softmax():
...
+@use_np
+def test_all():
+    A = np.ones((INT_OVERFLOW, 2))
+    A.attach_grad()
+    with mx.autograd.record():
+        B = np.all(A)
+    print(B)
+    assert B.asnumpy() == True
+    B.backward()
+    assert A.grad.shape == (INT_OVERFLOW, 2)
```
Comment (access2rohit) [inline tests/nightly/test_np_large_array.py:153]: bkwrd value check

File: tests/nightly/test_np_large_array.py [hunk 7]
```diff
@@ -76,3 +78,459 @@ def test_softmax():
...
+def test_amin():
+    A = np.ones((INT_OVERFLOW, 2))
+    A[100][1] = -1
+    A.attach_grad()
+    with mx.autograd.record():
+        B = np.amin(A)
+    print(B)
+    assert B.asnumpy() == -1.0
+    B.backward()
+    assert A.grad.shape == (INT_OVERFLOW, 2)
```
Comment (access2rohit) [inline tests/nightly/test_np_large_array.py:165]: bkwrd value check

File: tests/nightly/test_np_large_array.py [hunk 8]
```diff
@@ -76,3 +78,459 @@ def test_softmax():
...
+def test_amax():
+    A = np.zeros((INT_OVERFLOW, 2))
+    A[100][1] = 1
+    A.attach_grad()
+    with mx.autograd.record():
+        B = np.amax(A)
+    print(B)
+    assert B.asnumpy() == 1.0
+    B.backward()
+    assert A.grad.shape == (INT_OVERFLOW, 2)
```
Comment (access2rohit) [inline tests/nightly/test_np_large_array.py:177]: bkwrd value check

**Comment by Zha0q1** [comment]
The purpose of the tests is to check for int indexing. If an op uses index_t for indexing then it will pass the test. I don't see how going from small to large tensor will break correctness because indexing is not any part of the algorithm. Value check might make more sense for linalg operators because some linalg ops depend on indexing type to calculate pivots. But for sin, cos, min, those ops will just go through each value in the tensor

File: tests/nightly/test_np_large_array.py [hunk 9]
```diff
@@ -78,14 +79,669 @@ def test_softmax():
...
+    assert C.shape == (INT_OVERFLOW, 2)
+    assert C[0][0] == 2
+
+@use_np
+def test_all():
+    A = np.ones((INT_OVERFLOW, 2))
+    A.attach_grad()
+    with mx.autograd.record():
+        B = np.all(A)
+    assert B.asnumpy() == True
```
Comment (access2rohit) [inline tests/nightly/test_np_large_array.py:160]: question: even if these are numpy ops the output is not numpy compatible ?

Comment (Zha0q1) [reply tests/nightly/test_np_large_array.py:160]: It is numpy compatible. This is probably left over from my previous version. I removed `asnumpy()` locally and will probably include this fix in the next pr

**Comment by Zha0q1** [comment]
over night run:

================================= test session starts ==================================
platform linux -- Python 3.7.7, pytest-5.4.1, py-1.8.1, pluggy-0.13.1
rootdir: /home/ubuntu/incubator-mxnet, inifile: pytest.ini
plugins: remotedata-0.3.2, openfiles-0.4.0, astropy-header-0.1.2, hypothesis-5.8.3, arraydiff-0.3, doctestplus-0.5.0
collected 52 items                                                                     

tests/nightly/test_np_large_array.py .....s..........ss...s..........s...sss.... [ 82%]
...s.....                                                                        [100%]

=================================== warnings summary ===================================
tests/nightly/test_np_large_array.py:89
  /home/ubuntu/incubator-mxnet/tests/nightly/test_np_large_array.py:89: DeprecationWarning: invalid escape sequence \ 
    '''

tests/nightly/test_np_large_array.py:459
  /home/ubuntu/incubator-mxnet/tests/nightly/test_np_large_array.py:459: DeprecationWarning: invalid escape sequence \ 
    '''

-- Docs: https://docs.pytest.org/en/latest/warnings.html
====================== 43 passed, 9 skipped, 2 warnings in 1.10s =======================

File: tests/nightly/test_np_large_array.py (hunk 9 above)
Comment (access2rohit) [reply tests/nightly/test_np_large_array.py:160]: ok .... i will be expecting those changes in next PR then 👍

**Comment by access2rohit** [comment]
@mxnet-label-bot update [pr-ready-to-merge]

**Comment by access2rohit** [comment]
@mxnet-label-bot add [pr-ready-to-merge]

**Pull request merged**